EIPS/eip-2126.md

```diff
@@ -0,0 +1,143 @@
+
+## Motivation
+
+Currently Dapp developers are unable to identify the _type_ of a signature. By _type_ we are referring to how the signature was generated.
+
+If the Dapp sees the user's EOA (externally-owned address) as their account, then any signature will have been generated by the private key of that address, and the Dapp would easily be able to validate the signer of this signature as being the same account. On the other hand, if what a Dapp sees is a user's smart contract address (which is indirectly owned by the user's EOA) then obviously the Dapp would find a mismatch between signer address and account address.
+
+A real-world example would be using Metamask or [WalletConnect](https://walletconnect.org), where you could expose accounts which are either EOA accounts or smart contract accounts, which in turn would provide similar signatures but require different validation processes.
+
+The solution to this issue is to append a byte to the end of the signature which will correspond to one of the given signature types listed below. Dapps can then query this byte to know how to validate the given signature.
+
+## Specification
+
+### Formatting Signatures
+
+Example for [EIP-712](#EIP712) compliant signatures:
+
+```javascript
+const signatureBuffer = Buffer.concat([
+  ethUtil.toBuffer(ecSignatureRSV.v),
+  ethUtil.toBuffer(ecSignatureRSV.r),
+  ethUtil.toBuffer(ecSignatureRSV.s),
+  ethUtil.toBuffer(SignatureType.EIP712)
+]);
+```
+
+### Signature Types
+
+All signatures are represented as a byte array of arbitrary length, where the last byte (the "signature byte") specifies the signature type. The signature type is popped from the signature byte array before validation. The following signature types are supported within the protocol:
+
+| Signature byte | Signature type      |
+| -------------- | ------------------- |
+| 0x00           | [Illegal](#illegal) |
+| 0x01           | [Invalid](#invalid) |
+| 0x02           | [EIP712](#eip712)   |
+| 0x03           | [EthSign](#ethsign) |
+| 0x04           | [Wallet](#wallet)   |
+
+**Illegal**
+
+The is the default value of the signature byte. A transaction that includes an Illegal signature will be reverted. Therefore, users must explicitly specify a valid signature type.
+
+**Invalid**
+
+An `Invalid` signature always returns false. An invalid signature can always be recreated and is therefore offered explicitly. This signature type is largely used for testing purposes.
+
+**EIP712**
+
+An [`EIP712`](https://eips.ethereum.org/EIPS/eip-712) signature is considered valid if the address recovered from calling ecrecover with the given hash and decoded `v`, `r`, `s` values is the same as the specified signer. In this case, the signature is encoded in the following way:
+
+| Offset | Length | Contents            |
+| ------ | ------ | ------------------- |
+| 0x00   | 1      | v (always 27 or 28) |
+| 0x01   | 32     | r                   |
+| 0x21   | 32     | s                   |
```

Comment:
Usually wallet encode it r,s,v (see my comment https://ethereum-magicians.org/t/erc-2126-signature-type-recognition/3392/13?u=amxx)